Adds blame interaction graphs

boehmseb opened the pull request:
This PR adds a graph abstraction for blame interaction data with different views on the blame data.
Using graphs allows us to analyze blame interactions from a graph-theoretic perspective.

boehmseb committed:
Add BlameInteractionGraph class.

The class represents blame interaction data as a graph and provides different views on the data.

boehmseb committed:
Workspace change

boehmseb committed:
Fix the commit lookup function factory

boehmseb committed:
Finish node degree plot

boehmseb committed:
Implement arc plot

boehmseb committed:
Refactor chord plot

boehmseb committed:
Add degree scatter plot

boehmseb committed:
More useful nodes in arc plot

boehmseb committed:
Arc plot space edges

boehmseb committed:
Fix chord_plot_utils typing

boehmseb committed:
Fix more numpy related type errors

boehmseb committed:
Fix more type errors

boehmseb committed:
Rename blame interaction graph plots module

boehmseb committed:
Refactor blame interaction graph loading

boehmseb committed:
Reverse arc plot timeline

boehmseb committed:
change arc plot color scheme

boehmseb committed:
Merge branch 'f-BlameInteractionGraphs' of github.com:se-sic/VaRA-Tool-Suite into f-BlameInteractionGraphs

boehmseb committed:
Refactor commit lookup

boehmseb committed:
Fix bug in chord plot

boehmseb committed:
Initial commit interaction graph table

boehmseb committed:
Implement author interaction graph and refactor graph attribute types

boehmseb committed:
Merge branch 'f-BlameInteractionGraphs' of github.com:se-sic/VaRA-Tool-Suite into f-BlameInteractionGraphs

boehmseb committed:
Import TypedDict from typing_extensions if python < 3.8

boehmseb committed:
Fix element access on frozen set.

boehmseb committed:
Arc plot: remove arc endpoint spacing.

Unfortunately, this does not work well with plotly's zoom features.

boehmseb committed:
Add commit-author interaction graph

boehmseb committed:
Merge branch 'vara-dev' into f-BlameInteractionGraphs

boehmseb committed:
Add libgraphviz-dev to CI install requirements

boehmseb committed:
Drop python 3.6 and add 3.9 from CI

boehmseb committed:
Fix two broken tests

boehmseb committed:
Upgrade scikit-learn version requirement

cormensratio commented on varats/varats/plots/blame_interaction_graph_plots.py:
```diff
@@ -0,0 +1,428 @@
+      - sort: criteria to sort the revisions [degree, time]
+    """
+
+    NAME = 'cig_node_degrees'
```
Regarding the consistent naming of the prior ``NAME``s of the plots, shouldn't this be 'cig_node_degree_plot'? Or do you want to keep the NAME with a maximum of two underscores?

cormensratio commented on varats/varats/plots/blame_interaction_graph_plots.py:
```diff
@@ -0,0 +1,428 @@
+      - sort: criteria to sort the revisions [degree, time]
+    """
+
+    NAME = 'cig_node_degrees_scatter'
```
same as above

cormensratio commented on varats/varats/tables/blame_interaction_graph_table.py:
```diff
@@ -0,0 +1,111 @@
+class BlameInteractionGraphMetricsTable(Table):
+    """Visualizes bug metrics of a project."""
+
+    NAME = "cig_metrics_table"
```
Shouldn't it be called ``big_metrics_table``?

cormensratio reviewed (approved):
Besides my comments regarding the naming and after the passing of the pipeline checks, lgtm.

boehmseb committed:
Fix author interaction graph

boehmseb committed:
Fix commit-author interaction graph

boehmseb committed:
Simplify commit interaction graph table

boehmseb committed:
Merge branch 'vara-dev' into f-BlameInteractionGraphs

boehmseb committed:
BlameInteractionGraph views are no longer properties.

boehmseb committed:
Revert "Fix the commit lookup function factory"

This reverts commit 7cee70fb

boehmseb committed:
Fix bug in table artefact loading

boehmseb committed:
Clean up blame interaction graphs tables

boehmseb committed:
Fix cig_node_degrees plot

boehmseb committed:
Two alternative versions of node degree quantile plot

boehmseb committed:
Do not normalize axes in node degree quantile plot

boehmseb committed:
Add table showing commits with highest node degrees

boehmseb committed:
Add author interaction graph top degrees table

boehmseb committed:
Add author interaction graph node degree plot

boehmseb committed:
Add Commit-Author blame interaction plot/tables

boehmseb committed:
Fix bug in node degree scatter plot

boehmseb committed:
Merge branch 'vara-dev' into f-BlameInteractionGraphs

vulder commented on varats-core/varats/project/project_util.py:
```diff
@@ -66,10 +66,7 @@ def get_local_project_git_path(
-        Path,
-        Path(target_prefix()) / source.local.replace(os.sep, '-')
-    )
+    return tp.cast(Path, Path(target_prefix()) / source.local)
```
Just out of curiosity, why do we need a cast here? Should operator / on a Path not always return a Path?

vulder commented on varats/varats/data/reports/blame_interaction_graph.py:
```diff
@@ -163,7 +164,10 @@ def author_interaction_graph(self) -> nx.Graph:
-            return str(commit_lookup(u).author) == str(commit_lookup(v).author)
+            if u.commit_hash == DUMMY_COMMIT or v.commit_hash == DUMMY_COMMIT:
+                return u.commit_hash == v.commit_hash
+            return str(commit_lookup(u).author.name
+                      ) == str(commit_lookup(v).author.name)
```
I bet you checked this, but is `.name` not a string?

vulder commented on varats/varats/data/reports/blame_interaction_graph.py:
```diff
@@ -179,10 +183,14 @@ def edge_data(
             }
 
         def node_data(b: tp.Set[CommitRepoPair]) -> AIGNodeAttrs:
```
What is `b` here?

vulder commented on varats/varats/data/reports/blame_interaction_graph.py:
```diff
@@ -224,7 +231,9 @@ def commit_author_interaction_graph(
 
         commit_author_mapping = {
-            commit: commit_lookup(commit).author for commit in (list(ig.nodes))
+            commit: commit_lookup(commit).author.name
```
Here it seems to work without the extra `str()`

vulder commented on varats/varats/plots/blame_interaction_degree.py:
```diff
@@ -16,8 +16,8 @@
-from plotly import graph_objs as go
-from plotly import io as pio
+from plotly import graph_objs as go  # type: ignore
+from plotly import io as pio  # type: ignore
```
Shouldn't this be ignored by the plotly.* ignore in the config?

vulder reviewed (changes requested):
Just found a few small things and 2 questions.
Otherwise, this looks fine to me.

vulder commented:
A one thing I missed, try to update to vara-dev and check how the TC coverage is :wink: 
I like that it works now :laughing:

boehmseb committed:
Add code centrality plot and table

boehmseb committed:
Merge branch 'vara-dev' into f-BlameInteractionGraphs

vulder committed:
Add specific file bc file extension for blame data (#471)

J03D03 committed:
Update `vara-cs gen` example command (#466)

Use project instead of path to repository. Tiny change in paper.rst

cormensratio committed:
Adds a check for VaRA major releases (#431)

Adds a check for VaRA major releases when using vara-buildsetup vara -u or vara-buildsetup vara -b.

resolves se-sic/VaRA#565

vulder committed:
Implements upgrade functionallity for VaRA (#472)

resolves se-sic/VaRA#640

vulder committed:
Upgrade VaRA default version to 120 (#473)

boehmseb committed:
Better test isolation and config handling. (#476)

boehmseb committed:
add author information to tables

boehmseb committed:
Add commit-author box/violin plot

J03D03 committed:
Adds SVFPointsToAnalysisBenchmark as project (#463)

Adds SVFPointsToAnalysisBenchmark for testing alias analyses performance.

boehmseb committed:
Adds better SZZ support (#468)

Adds experiment to store BugProvider data in a report file and makes SZZ reports diffable by sorting the output.
Furthermore, adds a case study extender that uses bug fixing and introducing commits to initialize/extend a case study with bug data.

boehmseb committed:
Initial container support for varats (#422)

Adds initial support for varats to build base containers with research tools, so experiments can be run in isolation.

resolves se-sic/VaRA#718

boehmseb committed:
Fixes a bug where varats would crash when the BB config file does not exist (#478)

boehmseb committed:
Add scatterplots

boehmseb committed:
More plots

boehmseb committed:
Make interaction graph implementation reusable

boehmseb committed:
Implement file-based interaction graph

boehmseb committed:
Allow nodes to be indexed by commits or authors.

boehmseb committed:
Fix file filtering in file based interaction graph

boehmseb committed:
More data for author interaction graphs

boehmseb committed:
Implement __repr__() for CommitRepoPair

boehmseb committed:
Remove old unused plot

boehmseb committed:
Fix bug in file based interaction graph creation

boehmseb committed:
Add box plot to visualize diff between file and blame interactions

boehmseb committed:
Add table to visualize diff between file and blame interactions

boehmseb committed:
Merge branch 'vara-dev' into f-FileBasedAuthorInteractions

boehmseb committed:
Add graph cache helper

boehmseb committed:
Implement caching for interaction graphs

boehmseb committed:
Use git to calculate blame data instead of libgit2.

libgit2 is way too slow.

boehmseb committed:
Make get_local_project_git_path always return the correct path

boehmseb committed:
Fix type issues with build_cached_graph()

boehmseb committed:
Tune strip/violin plots

boehmseb committed:
Merge branch 'vara-dev' into f-FileBasedAuthorInteractions

boehmseb committed:
Merge branch 'vara-dev' into f-FileBasedAuthorInteractions

boehmseb committed:
Fix typing after merge

boehmseb committed:
Remove duplicate entry in mypy.ini

boehmseb committed:
Fix blame parsing

boehmseb committed:
Make violin plots flatter

boehmseb committed:
Initialize plot config from params

boehmseb committed:
Merge branch 'vara-dev' into f-FileBasedAuthorInteractions

boehmseb committed:
Fix bug in artefacts module

boehmseb committed:
Clean up blame interaction graph module

boehmseb committed:
Fix churn format for empty files

boehmseb committed:
Remove unwanted plots and tables

boehmseb committed:
Comment highlighting code

boehmseb committed:
Cleanup multivariate grid plot code

boehmseb committed:
Merge branch 'vara-dev' into f-FileBasedAuthorInteractions

boehmseb committed:
Add graphviz dependency to reviewdog

boehmseb committed:
Fix type issue with calc_commit_code_churn

boehmseb committed:
Make pylint (more) happy

boehmseb committed:
Refactor chord plot utils module

boehmseb committed:
Fix typo causing a type error

boehmseb committed:
Better annotations in central code scatter plot

boehmseb committed:
Fix kwargs key in build_plot

boehmseb committed:
Merge branch 'vara-dev' into f-BlameInteractionGraphs

boehmseb committed:
Fix revision lookup in cig chord plot

boehmseb committed:
Merge branch 'vara-dev' into f-BlameInteractionGraphs

boehmseb committed:
Fixes to cig chord and arc plots

boehmseb committed:
Merge branch 'vara-dev' into f-BlameInteractionGraphs

boehmseb committed:
Merge branch 'vara-dev' into f-FileBasedAuthorInteractions

boehmseb committed:
Merge branch 'f-FileBasedAuthorInteractions' into f-BlameInteractionGraphs

boehmseb committed:
Adapt to new plot architecture

boehmseb committed:
Upgrade numpy for better type hints.

boehmseb committed:
Fix some pylint warnings

boehmseb committed:
Merge branch 'vara-dev' into f-BlameInteractionGraphs

boehmseb committed:
Fix some typing issues

boehmseb committed:
Test get_submodule_head

boehmseb committed:
Test newest_processed_revision_for_case_study

boehmseb committed:
Test blame interaction graphs

boehmseb committed:
Fix report filter

boehmseb committed:
Test chord plot utils

boehmseb committed:
Move Tukey's fence implementation to metrics module

boehmseb committed:
Add missing path exists check in driver_artefacts.py

boehmseb committed:
Fix/adjust blame interaction scatter plots

boehmseb committed:
Merge branch 'vara-dev' into f-BlameInteractionGraphs

boehmseb committed:
Be python 3.7 compatible and make mypy happy

boehmseb committed:
Merge branch 'vara-dev' into f-BlameInteractionGraphs

boehmseb committed:
Fix doc formatting

boehmseb committed:
Merge branch 'vara-dev' into f-BlameInteractionGraphs

boehmseb committed:
Refactor

boehmseb committed:
Remove unused imports

boehmseb committed:
Remove caig_grid plot

vulder committed:
Merge branch 'vara-dev' into f-BlameInteractionGraphs

vulder commented on tests/utils/test_git_util.py:
```diff
@@ -80,39 +86,68 @@ def test_get_initial_commit_with_specified_path(self) -> None:
+        repo_head = FullCommitHash("cb15dfa4b2d7fba0d50e87b49f979c7f996b8ebc")
+        git_checkout(repo_head, repo_path)
+
+        try:
```
Why do you need a try here?

boehmseb replied:
Currently, we do not have a good way to isolate git repos in tests (using our test environment does a full clone each time).
Therefore, I need to ensure that I leave the repo in the state it was before the test. Hence the `finally` block with the checkout.

vulder commented on tests/utils/test_git_util.py:
```diff
@@ -80,39 +86,68 @@ def test_get_initial_commit_with_specified_path(self) -> None:
+        repo_head = FullCommitHash("cb15dfa4b2d7fba0d50e87b49f979c7f996b8ebc")
+        git_checkout(repo_head, repo_path)
+
+        try:
```
same here

vulder commented on tests/utils/test_git_util.py:
```diff
@@ -376,10 +411,10 @@ class TestCodeChurnCalculation(unittest.TestCase):
         """Check if we get the correct code churn for a single commit."""
 
-        repo = get_local_project_git("brotli")
+        repo = get_local_project_git_path("brotli")
```
above `repo` was rename to `repo_path` because you switched the func, we should keep this consistent

vulder commented on varats-core/varats/project/project_util.py:
```diff
@@ -87,10 +87,10 @@ def get_local_project_git_path(
+        raise AssertionError(f"Project {project_name} does not use git.")
 
-    return Path(target_prefix()) / Path(source.local)
+    return Path(source.fetch())
```
Fetch does return a Path tot the git? This is a strange API

boehmseb replied:
Yes it does.

vulder commented on varats-core/varats/project/project_util.py:
```diff
@@ -136,6 +136,28 @@ def get_local_project_git(
+    repos: tp.Dict[str, pygit2.Repository] = {}
+    project_cls = get_project_cls_by_name(project_name)
+
+    for source in project_cls.SOURCE:
+        source_name = os.path.basename(source.local)
+        repos[source_name] = get_local_project_git(project_name, source_name)
```
We should add a check here that the source is actually a git/gitsubmodule source and skip the others.

We'll have a lot of non git sources in the future.

vulder commented on varats-core/varats/utils/git_util.py:
```diff
@@ -172,6 +186,19 @@ def get_initial_commit(repo_folder: tp.Optional[Path] = None) -> FullCommitHash:
     )
 
 
+def git_checkout(
```
There should be a checkout method in the vara_manager.py, which uses the ProcessManager.

boehmseb replied:
There is, but it is framed to only be used for branches and needs to be adapted to also be used for commits.

On a side note: `vara_manager.py` probably isn't the best module for these kind of functions anymore.

vulder commented on varats-core/varats/utils/git_util.py:
```diff
@@ -542,6 +485,77 @@ def __hash__(self) -> int:
         return f"{self.repository_name}[{self.commit_hash}]"
 
+    def __repr__(self) -> str:
+        return self.__str__()
```
```suggestion
        return str(self)
```

vulder commented on varats-core/varats/utils/git_util.py:
```diff
@@ -554,7 +568,7 @@ def map_commits(
     # Skip 0000 hashes that we added to mark uncommitted files
     return [
-        func(commit_lookup(cr_pair.commit_hash, cr_pair.repository_name))
+        func(commit_lookup(cr_pair))
```
How does this work, without knowing the corresponding repository?

boehmseb replied:
It does know. All I do here is avoid unnecessarily unpacking the commit-repo-pair. The commit lookup function was refactored to do the unpacking by itself.

vulder commented on varats-core/varats/utils/git_util.py:
```diff
@@ -672,24 +686,42 @@ def calc_code_churn_range(
         (files changed, insertions, deletions)
     """
     churn_config = ChurnConfig.init_as_default_if_none(churn_config)
-    return calc_code_churn_range(repo, churn_config, commit, commit)[
```
Why do/can we no longer reuse range here?

boehmseb replied:
The range implementation does not work for merge commits as it displays diffs for multiple commits in that case but not the one you want. This implementation does not have that problem.
It  could be that you can merge these implementations if you find the right options but I do not know them.

vulder commented on varats/varats/data/metrics.py:
```diff
@@ -59,3 +61,68 @@ def normalized_gini_coefficient(distribution: pd.Series) -> float:
     return gini_coefficient(distribution) * (n / (n - 1.0))
+
+
+def apply_tukeys_fence(
```
Just checking, was there no tukeys_fence in scipy or some other lib ?

boehmseb replied:
Nope, just IQR calculations.

vulder commented on varats/varats/paper_mgmt/case_study.py:
```diff
@@ -73,6 +73,30 @@ class ExtenderStrategy(Enum):
     ADD_BUGS = 6
 
 
+def newest_processed_revision_for_case_study(
```
Do I understand this correctly: This function retrieves the newest revision (newest commit) that actually has results and, e.g., ignores newer revs that have no results up to now?

boehmseb replied:
Exactly

vulder commented on varats/varats/plots/blame_interaction_graph_plots.py:
```diff
@@ -0,0 +1,532 @@
+        commit = commit_lookup(node)
+        if not commit:
+            return False
+        # TODO: make filter configurable
```
If we don't need this, we should remove the TODO or?
Maybe just leave a comment with a hint: # Filter could be made configurable

boehmseb replied:
We didn't really use this plot yet except for some early visualizations, but it will only be useful if the filter is adjustable. However, I would postpone this to when we actually need the plot because then we will have a better understanding of what filters we need, hence, the TODO.

boehmseb committed:
Apply suggestions from code review

boehmseb committed:
Add test for code centrality table

boehmseb committed:
Fix typing

boehmseb committed:
Refactor and more tests for code churn

boehmseb committed:
Remove todo

boehmseb committed:
Test blame interaction graph tables

vulder committed:
Merge branch 'vara-dev' into f-BlameInteractionGraphs

vulder committed:
Merge branch 'vara-dev' into f-BlameInteractionGraphs

boehmseb commented on tests/table/test_blame_interaction_graph_table.py:
```diff
@@ -0,0 +1,130 @@
+{} & commits & authors & nodes & edges & \multicolumn{4}{c}{node degree} & \multicolumn{3}{c}{node out degree} & \multicolumn{3}{c}{node in degree} \\
+{} &        mean & median & min &  max &          median & min & max &         median & min & max \\
+\midrule
+\textbf{xz} &    1325 &      28 &   124 &   928 &       14.97 &    8.0 &   1 &  154 &             4.0 &   0 &  64 &            3.0 &   0 &  92 \\
```
```suggestion
\textbf{xz} &    1143 &      28 &   124 &   928 &       14.97 &    8.0 &   1 &  154 &             4.0 &   0 &  64 &            3.0 &   0 &  92 \\
```

boehmseb commented on tests/table/test_blame_interaction_graph_table.py:
```diff
@@ -0,0 +1,130 @@
+{} & commits & authors & nodes & edges & \multicolumn{4}{c}{node degree} & \multicolumn{3}{c}{node out degree} & \multicolumn{3}{c}{node in degree} \\
+{} &        mean & median & min & max &          median & min & max &         median & min & max \\
+\midrule
+\textbf{xz} &    1325 &      28 &     1 &     0 &         0.0 &    0.0 &   0 &   0 &             0.0 &   0 &   0 &            0.0 &   0 &   0 \\
```
```suggestion
\textbf{xz} &    1143 &      28 &     1 &     0 &         0.0 &    0.0 &   0 &   0 &             0.0 &   0 &   0 &            0.0 &   0 &   0 \\
```

boehmseb commented on tests/table/test_blame_interaction_graph_table.py:
```diff
@@ -0,0 +1,130 @@
+{} & commits & authors & nodes & edges & \multicolumn{4}{c}{node degree} & \multicolumn{3}{c}{node out degree} & \multicolumn{3}{c}{node in degree} \\
+{} &        mean & median & min & max &          median & min & max &         median & min & max \\
+\midrule
+\textbf{xz} &    1325 &      28 &   125 &    92 &        1.47 &    1.0 &   0 &  92 &             1.0 &   0 &   1 &            0.0 &   0 &  92 \\
```
```suggestion
\textbf{xz} &    1143 &      28 &   125 &    92 &        1.47 &    1.0 &   0 &  92 &             1.0 &   0 &   1 &            0.0 &   0 &  92 \\
```

boehmseb commented on varats/varats/tables/blame_interaction_graph_table.py:
```diff
@@ -0,0 +1,260 @@
+            pd.DataFrame.from_dict({
+                project_name: {
+                    ("commits", ""):
+                        int(project_git("rev-list", "--count", "HEAD")),
```
```suggestion
                        int(project_git("rev-list", "--count", revision.hash)),
```

boehmseb committed:
Fix flaky tests
